moveit_ros/trajectory_cache/src/features/get_cartesian_path_request_features.cpp

```diff
@@ -0,0 +1,498 @@
+  }
+
+  return moveit::core::MoveItErrorCode::SUCCESS;
+};
+
+MoveItErrorCode CartesianMaxStepAndJumpThresholdFeatures::appendFeaturesAsInsertMetadata(
+    Metadata& metadata, const GetCartesianPath::Request& source, const MoveGroupInterface& /*move_group*/) const
+{
+  metadata.append(name_ + ".max_step", source.max_step);
+
+  if (source.jump_threshold > 0)
+  {
+    metadata.append(name_ + ".jump_threshold", source.jump_threshold);
+  }
+  if (source.prismatic_jump_threshold > 0)
+  {
+    metadata.append(name_ + ".prismatic_jump_threshold", source.prismatic_jump_threshold);
+  }
+  if (source.revolute_jump_threshold > 0)
+  {
+    metadata.append(name_ + ".revolute_jump_threshold", source.revolute_jump_threshold);
+  }
+
+  return moveit::core::MoveItErrorCode::SUCCESS;
+};
+
+// CartesianWaypointsFeatures.
+
+CartesianWaypointsFeatures::CartesianWaypointsFeatures(double match_tolerance)
+  : name_("CartesianWaypointsFeatures"), match_tolerance_(match_tolerance)
+{
+}
+
+std::string CartesianWaypointsFeatures::getName() const
+{
+  return name_;
+}
+
+MoveItErrorCode CartesianWaypointsFeatures::appendFeaturesAsFuzzyFetchQuery(Query& query,
+                                                                            const GetCartesianPath::Request& source,
+                                                                            const MoveGroupInterface& move_group,
+                                                                            double exact_match_precision) const
+{
+  return appendFeaturesAsFetchQueryWithTolerance(query, source, move_group, match_tolerance_ + exact_match_precision);
+};
+
+MoveItErrorCode CartesianWaypointsFeatures::appendFeaturesAsExactFetchQuery(Query& query,
+                                                                            const GetCartesianPath::Request& source,
+                                                                            const MoveGroupInterface& move_group,
+                                                                            double exact_match_precision) const
+{
+  return appendFeaturesAsFetchQueryWithTolerance(query, source, move_group, exact_match_precision);
+};
+
+MoveItErrorCode CartesianWaypointsFeatures::appendFeaturesAsInsertMetadata(Metadata& metadata,
+                                                                           const GetCartesianPath::Request& source,
+                                                                           const MoveGroupInterface& move_group) const
+{
+  std::string path_request_frame_id = getCartesianPathRequestFrameId(move_group, source);
+  std::string base_frame = move_group.getRobotModel()->getModelFrame();
+
+  metadata.append(name_ + ".link_name", source.link_name);
+  metadata.append(name_ + ".robot_model.frame_id", base_frame);
+
+  // Waypoints.
+
+  // Restating them in terms of the robot model frame (usually base_link)
+  double x_offset = 0;
```

Comment:
This whole code block that comes up with identity transform offsets, then tries to do a TF lookup, etc. is repeated in a few places in this file. Consider factoring into a helper function.

Reply:
This is less trivial, but let me think of something.

Reply:
b55dfdee7baccc1893ba45cd69ce9cc60cfb87ed